### ISSUE TITLE
Problem Deploying SPSS.NET App Using ClickOnce

### ISSUE THREAD

**Issue by SydneyFixie** [AArnott/SPSS.NET#3]
Hi,

I've been using the SPSS.NET Wrapper for a couple of years now and appreciated the power of the wrapper. I have recently prepared a small WinForms data build app for one of my clients based on the current SPSS.NET repository using Visual Studio 2013. The app works fine inside Visual Studio but when I publish the app using ClickOnce and the CD-ROM option, any attempt to run it directly from my PC or the client's PC results in a system exception of "The Type Initiialiser for 'Spss.SpssThinWrapper' threw an exception. "

The Details dialog shows:

System.TypeInitializationException: The type initializer for 'Spss.SpssThinWrapper' threw an exception. ---> System.DllNotFoundException: C:\Users\Setup\AppData\Local\Apps\2.0\26XLD7NJ.AEO\9RYGRCCL.VKT\milt..tion_cdeb9d73f3f9454f_0001.0000_49d21d1c6226cfea\win64\spssio64.dll not found.
   at Spss.SpssThinWrapper.LoadSpssDll()
   at Spss.SpssThinWrapper..cctor()
   --- End of inner exception stack trace ---
   at Spss.SpssSafeWrapper.spssOpenRead(String fileName, Int32& handle)
   at Spss.SpssDataDocument..ctor(String filename, SpssFileAccess access)
   at Spss.SpssDataDocument.Open(String filename, SpssFileAccess access)
   at MiltonModels.FirstForm.button4_Click(Object sender, EventArgs e) ...<snip>

The problem relates to the Win32 and Win64 spssio*.dll and possibly the expected location of the DLLs in the app directory. When I had similar problems with an earlier version of the Wrapper I circumvented the problem by copying the DLLs into \Windows\System32\ and registering them with Regsvr32. 

I tried this approach and it did not work as I could not even register the DLLs (ie running regsvr32 spssio32.dll at a c:\ prompt in the \lib\win32\ directory gave an error "The module "spssio32.dll" was loaded but the entry-point DllRegisterServer was not found.Make sure that "spssio32.dll" is a valid DLL or OCX file and then try again.".) 

I then changed the Build Action on the SPSS dlls to "Embedded Resource" thinking that maybe it would bundle the DLLs in with the library but I still received the same LoadDll error. 

I note Microsoft does detail a manner to deploy COM components with CllickOnce (https://msdn.microsoft.com/en-us/library/ms165432(v=vs.120).aspx) but I have not been able recreate their approach using the SPSSIO DLLs. 

So in summary,

a) Is there an option I should be using in the ClickOnce deployment which will automatically bring the DLLs and dependencies to the install process on a new PC ?
b) If I can't use the ClickOnce approach, is there another way to build the app using Visual Studio that would allow me to easily install the app with the associated DLL dependencies?
c) Is there a solution/link that might show me how to use the Registration-Free COM approach?
d) Failing all that, how is it possible to register the DLLs on the client machine using RegSvr32 in a way that gets around the RegSvr32 problem identified above?

Any suggestions much appreciated,

Regards

Ian (Sydney is the non-de-plume)

**Comment by jduffman1319** [comment]
I'm having the same issue... any resolution to this problem?

**Comment by SydneyFixie** [comment]
Hi,

Would like to find a more elegant solution but in the end I lived with an ugly work around. Inside the LoadSpssDll routine inside SpssThinWrapper I removed the reference to the directory structure (found in the variable spssio32) and just loaded the spssio32.dll directly from the BaseDirectory. I also had to copy the relevant dll (32 bit in my case) to the root directory of the project, along with the support DLLs (icudt38.dll etc). The issue as best I can determine is that the Oneclick deployment doesn't bring the subdirectory structure into its seemingly arbitrarily named base directory after deployment. The main issue is that you can't switch easily between the 32 & 64 bit versions since the DLLs have the same name and you will need to copy only the version for the platform that you are targetting in the project's root directory. Happy to hear back if someone else has a superior solution.

Ian

**Comment by jduffman1319** [comment]
Would you mind sharing the revised spssio32.dll you created as well as the list of the supporting DLLs you mentioned (icudt38.dll, etc.)?  Thanks!

**Comment by AArnott** [comment]
If anyone cares about this at this point, let me know. I'm considering a revision to the nuget packaging to remove the subdirectory, but it would regress the "Any CPU" scenario, requiring an app to explicitly target x64 or x86.

Otherwise I'm going to close this as no one has talked on it for a few years.
